[Messenger] Define multiple buses from the `framework.messenger.buses` configuration

sroze opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #26652
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/issues/9617

Not everybody will benefit from having only one bus, especially with the CQRS-like usages. While keeping the extremely use of use of the single bus, this PR has the following:

- Create multiple buses from the YAML configuration
- Tag middleware only a specific buses
- Register middlewares from the YAML configuration

Even if it's visible in the PR's tests, here's how it will look like, for a completely full-customised version:
```yaml
framework:
    messenger:
        default_bus: commands
        buses:
            commands: ~
            events:
                middlewares:
                    - validation
                    - route_messages
                    - "Your\\Middleware\\Service"
                    - call_message_handler
```

A few things to note:
1. The YAML configuration creates `messenger.bus.[name]` services for the buses.
2. The YAML configuration for middleware just adds tags to the corresponding middlewares. 
3. If the middleware definition does not exists, it creates it. (without any magic on the arguments though, if it isn't auto-wirable, well... "your problem")
4. In the PR, there is this "TolerateNoHandler" middleware that is a great example for event buses

sroze committed:
[Messenger] Define multiple buses from the `framework.messenger.buses` configuration

ogizanagi commented:
IMHO, the `framework.messenger.buses.[bus_name].middlewares` node should be kept simple, with no priority handling (only the order matters). Otherwise, even if we'll have the debug command to help, it just brings confusion between services registered through the config and the ones through DI tags. It's far easier to read the list of middlewares applied on a bus on a single place.

I wonder if we shouldn't even remove the DI tag or just keep it for the default bus if `framework.messenger.buses.[default_bus_name].middlewares` is not set.

sroze commented:
> IMHO, the framework.messenger.buses.[bus_name].middlewares node should be kept simple, with no priority handling (only the order matters).

That's how I started it. The problem with this approach is how to place a middleware between/before/after the ones that are configured with the tags (and need priorities). I don't see any harm in being able to configure the priority, especially that when you don't, the configuration syntax is super easy. (i.e. by default, it's all middlewares priority 0, in the order in which you've put them in the configuration).

> Concerning attributes i'd say `bus: string|null` which either applies to the given or default bus. No big fan of the current explode trick compared to multiple tags.

It's actually convenient to do so within the code base to simplify the handling of either a middleware is only applied to a given set of buses or all of them, hence me doing this way here. I don't mind changing it but I don't have a strong argument in doing so tbh.

> Last but not least, i think st least 1 (default) bus should be configured.

That's the case if you don't configure your owns. If you do, then I'd argue it's now up to you :)

ro0NL commented:
> That's the case if you don't configure your owns.

Right, my bad. I missed the default bus entry in config :)

> I don't have a strong argument in doing so tbh

IMHO it's more pure :) but no strong opinion either.

About `framework.messenger.buses.[bus_name].middlewares`

You have to configure a service anyway, having to specify it here again is double bookkeeping, also this is not the place to infer new service definitions or so. IMHO :)

Having a single way of doing things is nice, and given you need to define the service i'd say tag it as a middleware as well, if needed.

ogizanagi commented:
> The problem with this approach is how to place a middleware between/before/after the ones that are configured with the tags (and need priorities).

That's why I'd suggest removing the tag 😄
The harm is not about having to handle a priority attribute in the config, but about allowing to configure things from 2 different places. I wouldn't allow both DI and semantic config for configuring middlewares.
I don't see much benefits to the DI tag for middlewares, while the semantic config is expressive and already allows to handle the stack order. That's how work the tactician-bundle. Never had to complain about it.

ro0NL commented:
A tag seems more flexible though :) considering third parties who want to register a middleware. Or put different what about the default middlewares currently?

ogizanagi commented:
> considering third parties who want to register a middleware.

That's the point to me: it's only a userland config. No third party implied. A third party could just help registering a default service to use, or provide a factory.

> Or put different what about the default middlewares currently?

They can just be configured as default values for the middlewares node.
If the user want to add a middleware before or after, they'll have to be explicit about it.

ro0NL commented:
Well.. i was considering exactly that, register/tag a middleware to to capture any message dispatched and pass it to a service.

sroze commented:
I think that your conversation is a good example of why we need to keep both configurations. 

@ogizanagi As far as I understand, your concern is that it can introduce some complexity about knowing where/how the middlewares are configuration: I appreciate that but I don't believe this will be the case as it is basically a question of how you, as a developer, use these configuration mechanisms. Though, if we see a large proportion of issues coming that are related to that, we can later decide which one to deprecate I guess.

romaricdrigon commented:
Hello all,
I would also be in favor of removing the tag. The duplicated config will be hard to trace, and can lead to side effect. 
Imagine I have an application 2 buses. Now I install a bundle that provides a middleware, as a tagged service, which I want to add to only one of the bus. How to deal with such situation? How to easily unregister the middleware from one bus?

ro0NL commented:
At this point im :+1: for keeping both configs (DI+semantic).

Please dont remove tagging middlewares, it makes it real hard to hook in and basically defeats the feature as decorating the bus at this point is probably easier :joy:

> Now I install a bundle that provides a middleware, as a tagged service, which I want to add to only one of the bus. How to deal with such situation? How to easily unregister the middleware from one bus?

It's probably the bundle's intend to enable the middleware on all buses. Otherwise it should allow you to configure the bus service to use.

sroze commented:
@ogizanagi What if you could enable/disable tags from the `framework.messenger.tags` or something? 🤔

sroze commented:
If we drop tags, "bundles" could use the prependConfiguration to have some sort of automated configuration.

ro0NL commented:
Im really curious what others think about @ogizanagi's statement:

> it's only a userland config. No third party implied

I tend to disagree with that. Why is a bus middleware "userland-config" and a kernel event listener not?

ogizanagi commented:
So, we've talked a bit on Slack with @sroze today. Here's mostly what I shared with him about my vision:

Here is how I see a typical configuration would look like with semantic config only and removing the DI tag:

```yaml
framework:
  messenger:
    buses:
      commands:
        middlewares:
          - logging # fwb alias
          - validation # fwb alias
          # DoctrineBundle transaction middleware.
          # This demonstrates support of factories we could provide, i.e the doctrine_transaction_middleware
          # is an abstract service the FrameworkExtension would decline with following arguments
          # (here "default" being the manager name to use. Only simple scalar arguments, i.e no services references)
          - doctrine_transaction_middleware: [default]
          # Third-parties would just have to provide a service id (or factories for most custom needs) and document usage.
          # No automatic middleware registering, the user opt-in explicitly for each bus.
          - msg_php.console_message_subscriber_middleware
          # userland middleware service id
          - app.messenger.my_custom_middleware
          - message_handler # alias
      events:
        middlewares:
          - logging
          - tolerate_no_handler # fwb alias
          - message_handler
```

where `logging`, `validation`, `tolerate_no_handler` & `message_handler` are aliases handled by the FrameworkBundle for convenience, which also remove by the same occasion the need for the `messenger.bus.middlewares.validation` keys from https://github.com/symfony/symfony/pull/26648.

Everything is configured in one place and explicit. Nothing behind the scene. No need for priority. No need for third-parties to register middlewares themself (just to provide and document middleware services ids/factories).

Supporting both config & DI tag means decentralizing the middleware stack, struggling with priorities (even if the debug command would help) and allowing third-parties to register middleware automagically (which I really don't think is useful).
A parrallel (limited though) could be made with Monolog's handlers stack: no DI tag is provided AFAIK to register handlers (good lord, thanks. That'll only bring more confusion :smile:).

Even the one legit use-case I see within the framework can be solved differently: the `MessengerDataCollector` has to be registered automatically on debug mode. The solution without tag is simple: just decorate the `HandleMessageMiddleware` just like we do with validator & event dispatcher collectors for instance.

Finally, I used Tactician for more than 2 years. [This is how the middlewares are registered by the bundle](https://github.com/thephpleague/tactician-bundle#configuring-middleware) (but simpler, no factory support). Never had to complain and I'm not aware of
any issue asking for a DI tag. So, what about starting with this?

Anyway, if a majority still prefers keeping the DI tag, then I'd be in favor of no semantic config at all. Supporting both just seems confusing and pointless to me. :)

ogizanagi commented:
To me, fwb known aliases would always win and are unlikely to collide with actual middlewares' ids, but yes, I guess a way to denote built-in aliases would be nice :)

chalasr commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1448,6 +1449,26 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
 
+        $container->setParameter('messenger.buses_names', array_keys($config['buses']));
+        foreach ($config['buses'] as $name => $bus) {
+            $container->setDefinition($busId = 'messenger.bus.'.$name, (new Definition(MessageBus::class, array(array())))->addTag('messenger.bus')->setPublic(true));
```
I'd put an abstract `messenger.bus` definition in xml to be extended here also

chalasr commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1448,6 +1449,26 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
+        foreach ($config['buses'] as $name => $bus) {
+            $container->setDefinition($busId = 'messenger.bus.'.$name, (new Definition(MessageBus::class, array(array())))->addTag('messenger.bus')->setPublic(true));
+
+            if ($name == $config['default_bus']) {
```
better use strict comparison

chalasr commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -162,4 +161,34 @@ private function registerReceivers(ContainerBuilder $container)
+        foreach ($container->findTaggedServiceIds('messenger.bus_middleware') as $serviceId => $tags) {
+            foreach ($tags as $tag) {
+                $priority = $tag['priority'] ?? 0;
+                $buses = isset($tag['bus']) ? explode(',', $tag['bus']) : $busesNames;
```
Not fond of the special syntax, AFAIK we are used to require adding as much tags as needed e.g.
```yaml
App\SomeMiddleware:
    tags:
        - { name: "messenger.bus_middleware", bus: "query" }
        - { name: "messenger.bus_middleware", bus: "command" }

chalasr reviewed (commented):
+1 for dropping the tag, not worth the confusion it could cause IMHO.

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1447,26 +1448,33 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
-                if (!$container->has($middleware['service'])) {
-                    $container->setDefinition($middleware['service'], new Definition($middleware['service']));
-                }
+            $container->setDefinition(
```
so in user-land you need a compiler pass, find all buses by tag (`messenger.bus`), and decorate each one.

or is it better to decorate `call_message_handler` once?

sroze replied:
I'd say it depends on your use-case. Both approaches would work.

sroze commented:
@ogizanagi @chalasr I've updated the PR. I agree with you, it's actually a better DX to have only one place for the configuration.

With the way the PR is right now, if you define you own middlewares, you override the default "logging", "call_message_handler" and "route_messages" handlers, so you **need** to define them. It feels like it could be a source of troubles for the users... What's your point of view on what we could/should do arround that?

ogizanagi commented:
> you override the default "logging", "call_message_handler" and "route_messages" handlers, so you need to define them.

I don't think it's an issue. The only thing we should _maybe_ do is hint the user if the `call_message_handler` is not part of the list. Or automatically adding it but that would be a bit strange to treat it differently from other middleware.

ogizanagi commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml:
```diff
@@ -51,16 +38,14 @@
 
         <!-- Logging & Debug -->
-        <service id="messenger.middleware.debug.logging" class="Symfony\Component\Messenger\Debug\LoggingMiddleware">
+        <service id="messenger.middleware.logging" class="Symfony\Component\Messenger\Middleware\LoggingMiddleware">
```
IMHO, each middleware should be abstract and declined for each bus in order to be able to decorate them independently if needed.

sroze replied:
What's the use case of decorating a middleware?

ogizanagi replied:
There could be many. The one I may suggest after this one is an `ActivationMiddlewareDecorator`, that will decorate a middleware to conditionally enable it. For instance, in debug mode only. That could solve https://github.com/symfony/symfony/issues/26901 differently.

ogizanagi commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -1038,6 +1039,32 @@ function ($a) {
+                                            ->end()
+                                            ->children()
+                                                ->scalarNode('service')->isRequired()->end()
+                                                ->integerNode('priority')->defaultValue(0)->end()
```
can be removed now? 😄The `service` key wouldn't be necessary anymore neither so it can be simplified.

sroze commented:
> The only thing we should maybe do is hint the user if the call_message_handler is not part of

I think we should even throw an exception. Worst case, they really don't want a handler to be called, they use a middleware like the one proposed in #26901

sroze commented:
Rebased. Ready for more reviews @symfony/deciders.

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1446,6 +1448,33 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
+
+            $container->setDefinition($busId = 'messenger.bus.'.$name, (new Definition(MessageBus::class, array($middlewares)))->addTag('messenger.bus'));
+
+            if ($name === $config['default_bus']) {
```
dx-wise i'd expect an exception in case of an unknown bus is configured to be the default

ro0NL commented on src/Symfony/Component/Messenger/DataCollector/MessengerDataCollector.php:
```diff
@@ -48,9 +47,9 @@ public function reset()
+    public function collectMessage($message, $result = null, \Throwable $exception = null)
     {
         $debugRepresentation = array(
             'message' => array(
```
should we expose the bus a message is dispatched by?

ro0NL commented on src/Symfony/Component/Messenger/DataCollector/BusCollector.php:
```diff
@@ -0,0 +1,44 @@
+    private $decoratedBus;
+    private $collector;
+
+    public function __construct(MessageBusInterface $decoratedBus, MessengerDataCollector $collector)
```
why not a traceable bus implementation instead and collect from the data collector, how it's usually done AFAIK :)

ro0NL commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -45,14 +43,21 @@ public function __construct(string $messageBusService = 'message_bus', string $m
             $container->removeDefinition('messenger.middleware.debug.logging');
         }
 
+        $buses = $container->getParameter('messenger.buses');
```
find buses by tag? so no intermediary param is needed...

ro0NL commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -45,14 +43,21 @@ public function __construct(string $messageBusService = 'message_bus', string $m
     public function process(ContainerBuilder $container)
     {
-        if (!$container->hasDefinition($this->messageBusService)) {
+        if (!$container->hasDefinition('messenger.handler_resolver')) {
```
`$this-> $messageHandlerResolverService`?

sroze commented:
@ro0NL @ogizanagi thank you for your reviews. I've updated the PR the following way:
1. We create middlewares for each bus, based on a `parent` service.
2. The collector is registered some "traceable" bus, same as we do for the cache & events
3. Removed the `buses` parameters to only rely on the `message_bus` tag

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php:
```diff
@@ -569,6 +569,19 @@ public function testMessengerAdapter()
+        $this->assertEquals(array(new Reference('messenger.bus.events.middleware.route_messages'), new Reference('messenger.bus.events.middleware.'.TolerateNoHandler::class), new Reference('messenger.bus.events.middleware.call_message_handler')), $container->getDefinition('messenger.bus.events')->getArgument(0));
+
+        $this->assertTrue($container->hasAlias('message_bus'));
+        $this->assertEquals('messenger.bus.commands', (string) $container->getAlias('message_bus'));
```
`assertSame`

ro0NL commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -45,12 +43,21 @@ public function __construct(string $messageBusService = 'message_bus', string $m
-        if (!$container->getParameter('kernel.debug') || !$container->hasAlias('logger')) {
-            $container->removeDefinition('messenger.middleware.debug.logging');
+        $collectorDefinition = $container->getDefinition('messenger.data_collector');
+        foreach ($container->findTaggedServiceIds('messenger.bus') as $busId => $tags) {
```
for consistency maybe `$this->busTag`, or at this point _just_ inline magic strings =/

no real opinion :)

sroze replied:
I changed it to be consistent and only inject the tag names as the configuration; it makes more sense.

ro0NL commented on src/Symfony/Component/Messenger/Middleware/TolerateNoHandler.php:
```diff
@@ -0,0 +1,30 @@
+/**
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+class TolerateNoHandler implements MiddlewareInterface
```
perhaps a little unit test for it :)

sroze replied:
Yep, added tests for these two classes 😉

ro0NL commented on src/Symfony/Component/Messenger/TraceableMessageBus.php:
```diff
@@ -0,0 +1,55 @@
+/**
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+class TraceableMessageBus implements MessageBusInterface
```
same, unit testing low hanging fruit is nice :)

ro0NL reviewed (approved):
real nice :+1:

sroze commented:
And rebased 👍

chalasr commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1461,6 +1463,36 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
+            }
+
+            $container->setDefinition(
+                $busId.'.trace',
```
the convention is `debug.$id`

sroze replied:
`debug.traced.$id` ?

chalasr replied:
why not :)

sroze replied:
Updated to "follow the convention" 😉

chalasr reviewed (approved):
1 minor comment 👍 
doc issue created https://github.com/symfony/symfony-docs/issues/9617

chalasr commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -29,28 +29,39 @@ class MessengerPass implements CompilerPassInterface
 
-        if (!$container->getParameter('kernel.debug') || !$container->hasAlias('logger')) {
-            $container->removeDefinition('messenger.middleware.debug.logging');
+        $collectorDefinition = $container->getDefinition('messenger.data_collector');
```
kind of couples the pass to WebProfiler (basically Framework), I would skip if the definition does not exist

sroze replied:
Fair enough.

sroze commented:
You are right, my example in the original pull-request description wasn't accurate anymore (updated now). The reason you had this exception is because it tried to create an instance of the middleware for this bus based on the parent definition that you define. 

In the example you mentioned, you could even:
```patch
- 'messenger.middleware.tolerate_no_handler'
+ tolerate_no_handler
```

Regarding the `buses.[name].type` configuration, I like the idea but is obviously trickier to settle on what should be in which. Therefore I'd argue this shouldn't be in this PR at least 🙃 (could be a good 3rd party bundle as well).

weaverryan commented:
Hey guys!

Sorry to come late after so much work. 

> With the way the PR is right now, if you define you own middlewares, you override the default "logging", "call_message_handler" and "route_messages" handlers, so you need to define them. It feels like it could be a source of troubles for the users... What's your point of view on what we could/should do around that?

I actually think this is a *huge* problem. Imagine a user wants to add a new middleware: they setup their config, set middleware to `[my_new_middleware]` and suddenly things that worked before don't - my message isn't being handled or routed.

I'm not sure of the best solution - I think the argument around tags is valid, and the fact that experience with TacticianBundle has validated their approach a bit. Actually, one solution would be to be *more* like TacticianBundle: have *zero* default middleware out-of-the-box. Then, in the recipe, include the 3 middleware in `messenger.yaml` with comments. When a user creates a second bus, they'll copy and paste.

weaverryan commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -1052,6 +1053,21 @@ function ($a) {
                                 ->end()
                             ->end()
                         ->end()
+                        ->scalarNode('default_bus')->defaultValue('default')->end()
```
Could we default this to null, then in the extension:

1) If there is 1 extension, just use it
2) If there are multiple, throw a clear exception that the `default_bus` key needs to be defined?

I could see a situation where someone sees:

> The default bus named "default" is not defined. Define it or change the default bus name.

And thinking, where did default come from?

sroze replied:
Yep, I like that idea 👍

sroze replied:
Changed it this way.

sroze commented:
> Actually, one solution would be to be more like TacticianBundle: have zero default middleware out-of-the-box.

This means we can't auto-activate the messenger when we know it's installed (i.e. the `MessageBus` class exists) and we would only rely on the YAML configuration. It's pretty good to be able to do this (i.e. doing some configuration via the recipe) but it's more boilerplate (and much harder to maintain, once the recipe is installed, you're done...)

What about having the following`middlewares` per bus:
- `logging`
- _all the middlewares in the `buses.[name].middlewares` configuration_
- `route_messages`
- `call_message_handler`

As a developer, you can disable the "default middlewares" with:
```yaml
buses.[name].default_middlewares: false
```

sroze commented:
I've pushed 3 commits:
1. [`0d78a0e`](https://github.com/symfony/symfony/pull/26864/commits/0d78a0e7683280ccb9e82429e641915cb5836794) that allows middlewares to be services (no requirement for them to be abstract)
2. [` d39604e`](https://github.com/symfony/symfony/pull/26864/commits/d39604e634bd4f3f08103422a0a4b75c068f9daa) forces to have a matching `default_bus` only when using more than one bus. Otherwise, well, it take the name of that only bus.
3. [`8d16cda`](https://github.com/symfony/symfony/pull/26864/commits/8d16cdabc8e3790a6c9ab306db39f9a4488f3114) my favourite alternative to replace all the configured middlewares: we have a set of hard-coded default middlewares and you can squeeze yours in between. If you really need to (I believe this will be rare) you can disable the default middlewares with `default_middlewares: false`.

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1461,21 +1463,54 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
+        }
 
+        $defaultMiddlewares = array('before' => array('logging'), 'after' => array('route_messages', 'call_message_handler'));
         if ($config['middlewares']['validation']['enabled']) {
```
at this point..shouldnt we let users this configure as a regular middleware, per bus, if needed? Extra configs feels a bit too much config IMHO. Or try to prepend either way like 'logging'.

ro0NL replied:
Or put different, it's a confusing config as it relies on `default_middlewares` also.

weaverryan replied:
This *one* special middleware handling is indeed a little bit weird. Is there any special reason when you would *not* want this middleware? Could it ever cause issues? My instinct is to *always* include it when `default_middlewares` is enabled. If you *really* don't want it for some reason, you can set that config to false (unless there's a somewhat common use-case for not wanting this middleware).

ro0NL replied:
I tend to opt-in to a validation middleware.. im not sure validating messages by default makes sense.

Also it's not like adding `- validation` is "bad config" or so =/

sroze replied:
> Is there any special reason when you would not want this middleware?

Yes, when the component is not included.

> I tend to opt-in to a validation middleware.. im not sure validating messages by default makes sense.

Let's do this. So it is explicit and removes some extra configuration in there.

weaverryan reviewed (commented):
Other than one small comment, yea, I like this approach. I can't think of any other part of the framework that has this exact problem, and so, the solution is also a bit unique. But, it's both practical (works immediately for the 90%) and extensible.

weaverryan commented:
Ping @symfony/deciders! 

I'm 👍 for this approach, and it's a big PR, so we need to get it merged in soon. The most important aspect is how middleware are added. The final decision about that is explained very well here: https://github.com/symfony/symfony/pull/26864#issuecomment-383301982

ro0NL commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -177,4 +188,42 @@ private function registerSenders(ContainerBuilder $container)
+            }
+
+            if ('messenger.middleware.validation' === $messengerMiddlewareId && !$container->has('validator')) {
+                throw new RuntimeException('The Validation middleware is only available when the Validator component is installed and enabled. Try running "composer require symfony/validator".');
```
would be nice if we can improve such error messages at the DI service level

sroze replied:
I think it's a wider question within core. That's the best way I found to keep the great DX while using this middleware without the Validator component. I would keep it here.

ro0NL replied:
See #27004

ogizanagi replied:
IMHO, this still belongs to the Fwb extension where we have all the keys to detect if validation is installed and enabled. The `messenger.middleware.validation` is really specific to the Fwb wiring, so the pass alone doesn't have to deal with this unless strictly necessary.

sroze replied:
I see the point that this has little to do in here. Moved to the FrameworkBundle 👍

ro0NL commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -29,28 +30,38 @@ class MessengerPass implements CompilerPassInterface
+            }
+
+            if ($container->hasDefinition('messenger.data_collector')) {
+                $this->registerBusToCollector($container, $busId, $tags[0]);
```
should we iterate `$tags` as well or throw in case of >1?

sroze replied:
I don't believe there is a use-case for a bus to be tagged multiple times 🤔

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1461,21 +1463,44 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
+                $container->setAlias(MessageBusInterface::class, $busId);
+            }
+
+            $container->setDefinition(
```
should/can we process this at the compile pass level?

sroze replied:
Yes, very good idea! So we add it only when we need it 💃

ro0NL reviewed (commented):
now we have the `messenger.bus` tag instead of tagging middlewares.. from a vendor perspective im leaning to wire my own bus, thus tag it. Given that i think it should work like any other bus that comes from config.

There's no framework-way of asking for an event or command bus, and i want to avoid configuration for that on my side. So i'll autowire the default bus to be used for commands, but then i still need one for events. Alternatively i could wire a one-size-fits-all bus by injecting all tagged buses, and break on first successful bus handling a message.

ogizanagi commented on src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/php/messenger_multiple_buses.php:
```diff
@@ -0,0 +1,23 @@
+                'default_middlewares' => false,
+                'middlewares' => array(
+                    'route_messages',
+                    'Symfony\\Component\\Messenger\\Middleware\\TolerateNoHandler',
```
It's a bit misleading reading the tests and seing this FQCN while the actual middleware exists and has a service id that should be used (or the shortcut), not the FQCN. Is it really relevant to test a FQCN here?

sroze replied:
Good point; changed.

ogizanagi commented on src/Symfony/Component/Messenger/DataCollector/MessengerDataCollector.php:
```diff
@@ -14,21 +14,34 @@
 {
+    private $traceableBuses = array();
+
+    public function registerBus(string $name, TraceableMessageBus $bus)
```
Why not injected in the constructor?

sroze replied:
This is to match with what already exists like the [`CacheDataCollector`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/DataCollector/CacheDataCollector.php#L36-L39) for example.

ogizanagi commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -29,28 +31,38 @@ class MessengerPass implements CompilerPassInterface
+                $container->getParameterBag()->remove($busMiddlewaresParameter);
+            }
+
+            if ($container->getParameter('kernel.debug') && $container->hasDefinition('messenger.data_collector')) {
```
Is the `$container->getParameter('kernel.debug')` check useful, or `$container->hasDefinition('messenger.data_collector')` would be enough?

sroze replied:
True, we can remove the `$container->getParameter('kernel.debug')` now.

ogizanagi commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1461,21 +1462,39 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
+            $config['default_bus'] = array_keys($config['buses'])[0];
+        }
+
+        $defaultMiddlewares = array('before' => array('logging'), 'after' => array('route_messages', 'call_message_handler'));
```
Isn't `validation` supposed to be in `defaultMiddlewares['before']` if the component is enabled, as described in https://github.com/symfony/symfony/pull/26864#issuecomment-383301982 ?

ro0NL replied:
See https://github.com/symfony/symfony/pull/26864#discussion_r183215246

ogizanagi replied:
Alright, let's stick with it :)

ogizanagi reviewed (approved):
👍 With only minor comments. Thanks for considering what we discussed on Slack.
I'm not found of the default middlewares vs always using explicit configuration but I guess that's a good tradeoff.

sroze commented:
AppVeyor's failure is un-related :)

sroze commented:
Repushed the last commit to make AppVeyor green again 💚

sroze commented:
Let's merge! Thank you all for your feedback 🎉
